feat: ":write!" skips "file changed" warning

lorlouis opened the pull request:
Small fix for https://github.com/neovim/neovim/issues/7270

I also added some tests to prevent regression.

gpanders commented:
Related https://github.com/neovim/neovim/issues/18290

This is probably up to @justinmk whether this is the approach we want to take (among the alternatives listed in #7270). Seems reasonable to me though.

justinmk commented on test/functional/ui/input_spec.lua:
```diff
@@ -406,6 +406,78 @@ describe('input non-printable chars', function()
+
+  it("does not ask for confirmation when force write", function()
+    write_file("Xtest-overwrite-forced", [[foobar]])
+    local screen = Screen.new(60,8)
```
tests can be much less verbose by using a smaller screen where there's not much need for a big screen.
```suggestion
    local screen = Screen.new(40,4)
```

justinmk reviewed (commented):
needs a doc update (probably in a few places), then I guess we could try this. 

Ideally we would show the warning but just ensure that there's no hit-enter prompt. Could be done by doing a redraw before showing the message.

justinmk commented on test/functional/ui/input_spec.lua:
```diff
@@ -406,6 +406,61 @@ describe('input non-printable chars', function()
   end)
 end)
 
+describe('skip do you really want to write on force write', function()
```
input_spec is not the correct place for this test.

justinmk commented on test/functional/ui/input_spec.lua:
```diff
@@ -406,6 +406,61 @@ describe('input non-printable chars', function()
+    os.remove('Xtest-overwrite-forced')
+  end)
+
+  it("does not ask for confirmation when force write", function()
```
```suggestion
  it('w! does not show "file has been changed" warning', function()
```

justinmk commented on test/functional/ui/input_spec.lua:
```diff
@@ -406,6 +406,61 @@ describe('input non-printable chars', function()
+    ]])
+
+    -- The timestamp is in second resolution, wait two seconds to be sure.
+    screen:sleep(2000)
```
The test could be a lot faster by manually changing the time via `utime`:

https://github.com/luvit/luv/blob/master/docs.md#uvfs_utimepath-atime-mtime-callback

    helpers.nvim_exec_lua([[vim.loop.fs_utime('Xtest-overwrite-forced', 999999999, 999999999)]])

This is worth the trouble to save 2 seconds. If all of our tests took 1 second, the test suite would take 66 hours to run.

lorlouis replied:
I did not used this exact line, but It should achieve the same result.

justinmk commented on test/functional/ui/input_spec.lua:
```diff
@@ -406,6 +406,61 @@ describe('input non-printable chars', function()
+      {1:~                                       }|
+                                              |
+    ]])
+    feed("n")
```
would use `screen:expect_unchanged()` here. But really it's not necessary.
Instead use command() because it's synchronous (whereas feed() is async), then
check v:errmsg.

```
command('let v:errmsg = ""')
command('w!')
eq('', eval('v:errmsg'))
```

lorlouis replied:
I accidentally called feed() instead of feed_command(). When calling the proper method, the screens differ.

justinmk replied:
if possible, use `command()` (synchronous and error-checked) instead `feed_command()` (asyncrhonous, errors are not surfaced in the test harness)

lorlouis replied:
I made the changes, but I had to use `feed_command("w")` when testing that the prompt appears otherwise the test would just hang, I added a comment.

justinmk commented on test/functional/ui/input_spec.lua:
```diff
@@ -406,6 +406,61 @@ describe('input non-printable chars', function()
+    ]])
+
+    feed("w!")
+    screen:expect([[
```
```suggestion
    -- Use a screen test because the warning does not set v:errmsg.
    screen:expect([[
```

justinmk commented on test/functional/ui/input_spec.lua:
```diff
@@ -411,7 +411,7 @@ describe('skip do you really want to write on force write', function()
   end)
 
-  it("does not ask for confirmation when force write", function()
+  it("w! does not show \"file has been changed\" warning", function()
```
just use single quotes. don't do this escaping stuff in descriptions.
```suggestion
  it(':w! does not show "file has been changed" warning', function()
```

lorlouis committed:
fix: w! showing a warning when it should not

superficial fixes

added an entry in vim_diff.txt

fixed vim_diff.txt wrong message

Update runtime/doc/vim_diff.txt

Co-authored-by: zeertzjq <zeertzjq@outlook.com>

moved an os.remove into after_each block

lorlouis commented:
Sorry for the spam, I messed up a rebase on my end.

lorlouis commented on test/functional/core/fileio_spec.lua:
```diff
@@ -139,3 +144,60 @@ describe('fileio', function()
+  end)
+
+  it(':w! does not show "file has been changed" warning', function()
+    write_file("Xtest-overwrite-forced", [[foobar]])
```
I made the change, thanks for the suggestion.

zeertzjq commented on test/functional/core/fileio_spec.lua:
```diff
@@ -139,3 +144,61 @@ describe('fileio', function()
+  end)
+
+  it(':w! does not show "file has been changed" warning', function()
+    write_file("Xtest-overwrite-forced", 'foobar\n')
```
Hmm, unfortunately test still fails on Windows. Then I have another suggestion
```suggestion
    write_file("Xtest-overwrite-forced", 'foobar')
    command('set nofixendofline')
```

lorlouis replied:
I do not have access to a windows machine to test this out but I think it might be related to the lack of a file extension. I'll commit that first and if the ci still fails after that I'll try your patch.

dundargoc commented:
Rebase please.

lorlouis commented:
Right now master seems broken, when I try to run the tests I get a `vim/_init_packages.lua:0: module 'vim.shared' not found` which is the same error this branch currently generates. I don't really have the time to fix it but once it gets fixed I'll be able to make sure the tests pass.

zeertzjq commented on test/functional/core/fileio_spec.lua:
```diff
@@ -2,6 +2,7 @@ local helpers = require('test.functional.helpers')(after_each)
 
 local assert_log = helpers.assert_log
 local assert_nolog = helpers.assert_nolog
+local write_file = helpers.write_file
```
This variable is defined again below, so this line should be removed.

lorlouis replied:
thanks

zeertzjq commented on test/functional/core/fileio_spec.lua:
```diff
@@ -212,3 +215,63 @@ describe('tmpdir', function()
   end)
 end)
+
+describe('skip do you really want to write on force write', function()
```
Missing `before_each`
```suggestion
describe('skip do you really want to write on force write', function()
  before_each(clear)
```

clason commented:
@justinmk ping -- this was marked for 0.8; is this still on the table?

zeertzjq commented:
Looks fine to me, but I'm not going to approve.

justinmk commented on test/functional/core/fileio_spec.lua:
```diff
@@ -219,3 +221,64 @@ describe('tmpdir', function()
   end)
 end)
+
+describe('skip do you really want to write on force write', function()
```
there is no need for a separate describe(), this should live in the describe() starting at line 30

justinmk commented on src/nvim/fileio.c:
```diff
@@ -2534,8 +2534,9 @@ int buf_write(buf_T *buf, char *fname, char *sfname, linenr_T start, linenr_T en
-    // Check if the timestamp hasn't changed since reading the file.
-    if (overwriting) {
+    // If 'forceit' is false,
+    // check if the timestamp hasn't changed since reading the file.
```
comment can fit on one line.

justinmk commented:
needs a mention in vim_diff.txt. Remaining comments are trivial to fix, then ready to merge. @lorlouis do you have time today for this?

lorlouis commented:
> needs a mention in vim_diff.txt. Remaining comments are trivial to fix, then ready to merge. @lorlouis do you have time today for this?

Just woke up, I'll fix the pr in a few hours.

zeertzjq commented on runtime/doc/vim_diff.txt:
```diff
@@ -372,6 +372,7 @@ Lua interface (|lua.txt|):
 Commands:
   |:doautocmd| does not warn about "No matching autocommands".
   |:wincmd| accepts a count.
+  |:write| shows a prompt if the file was updated externally.
```
This is wrong.

lorlouis replied:
You're right,  I changed the message. Not sure if it's the right place to put it though.

lorlouis commented:
There seem to be a permission problem with libuv in the ci, not sure what to do with that.

zeertzjq commented on runtime/doc/vim_diff.txt:
```diff
@@ -372,6 +372,7 @@ Lua interface (|lua.txt|):
 Commands:
   |:doautocmd| does not warn about "No matching autocommands".
   |:wincmd| accepts a count.
+  |:write!| does not show a prompt if the file was updated externally.
```
```suggestion
  `:write!` does not show a prompt if the file was updated externally.
```

justinmk commented:
> There seem to be a permission problem with libuv in the ci, not sure what to do with that.

If you are referring to this:

> ERR 2022-09-24T20:37:54.170 T5450.36069.0 libuv_process_spawn:84: uv_spawn(fake_shell) failed: no such file or directory

That's just a log message captured from test T5450 (notice the "TXX" test-id in the message). Logs are printed if a test fails.

The actual test failure is an invalid help tag, as @zeertzjq noted.

justinmk commented on test/functional/core/fileio_spec.lua:
```diff
@@ -151,6 +153,62 @@ describe('fileio', function()
+      {1:~                                       }|
+      <erwrite-forced" [noeol] 1L, 6B written |
+    ]])
+    os.remove('Xtest-overwrite-forced')
```
this can go in the after_each() block

ben-albrecht commented:
Was this included as part of the Nvim v0.8.0 release? I am trying to see if my local version has this change, but didn't see this mentioned in release notes.

dundargoc commented:
I don't think so, I don't see a backport label.

lorlouis commented:
@dundargoc the changes are present in the commit used to build 8.0.0
https://github.com/neovim/neovim/blob/d367ed9b23d481998d297d812f54b950e5511c24/src/nvim/fileio.c#L2540
